# 🎓 **Итоговая работа. Анализ кинокассовых сборов**

**Датасет:** [TMDB 5000 Movie Dataset](https://www.kaggle.com/datasets/tmdb/tmdb-movie-metadata)

---

Вы — аналитик в крупной киностудии. Руководство планирует новый фильм и хочет принять решение о бюджете, жанре и стратегии релиза на основе данных, а не интуиции.

В вашем распоряжении — база данных **TMDB** (The Movie Database) с информацией о тысячах фильмов: бюджеты, кассовые сборы, оценки зрителей, жанры и другие характеристики.

## 🎯 Задачи

Работа состоит из **трёх частей**:

| Часть | Тема | Вес |
|-------|------|-----|
| **1. EDA** | Исследовательский анализ данных | ~40% |
| **2. Корреляция** | Анализ взаимосвязей между признаками | ~25% |
| **3. Гипотезы** | Проверка статистических гипотез | ~35% |

## 📊 Описание датасета

**Ключевые признаки:**

| Признак | Описание |
|---------|----------|
| `budget` | Бюджет фильма (USD) |
| `revenue` | Мировые кассовые сборы (USD) |
| `vote_average` | Средняя оценка зрителей (0–10) |
| `vote_count` | Количество голосов |
| `runtime` | Длительность (минуты) |
| `release_date` | Дата выхода |
| `genres` | Жанры (строка в формате JSON) |
| `original_language` | Язык оригинала (код: ``en``, ``fr``, ``ja``, ...) |
| `popularity` | Индекс популярности TMDB |
| `title` | Название фильма |
| `tagline` | Слоган фильма |
| `overview` | Краткое описание |
| `homepage` | Сайт фильма |
| `status` | Статус (Released, Post Production, ...) |

## ✏️ Требования к оформлению графиков

- У каждого графика должен быть **заголовок** (``plt.title()``)
- Подписаны **оси** (``plt.xlabel()``, ``plt.ylabel()``)
- Легенда — при наличии нескольких элементов
- Оставляйте **комментарии и выводы** к каждому разделу


---
# **Часть 1. Исследовательский анализ данных (EDA)**
---

## **1. Импорт библиотек и загрузка данных**

Импортируйте библиотеки и загрузите датасет.

In [52]:
import json

import kagglehub
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from scipy import stats

# Загрузка данных
path = kagglehub.dataset_download("tmdb/tmdb-movie-metadata")
df = pd.read_csv(f"{path}/tmdb_5000_movies.csv")
df.head()

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...","[{""iso_3166_1"": ""GB"", ""name"": ""United Kingdom""...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466
3,250000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 80, ""nam...",http://www.thedarkknightrises.com/,49026,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",en,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,"[{""name"": ""Legendary Pictures"", ""id"": 923}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-07-16,1084939099,165.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The Legend Ends,The Dark Knight Rises,7.6,9106
4,260000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://movies.disney.com/john-carter,49529,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...",en,John Carter,"John Carter is a war-weary, former military ca...",43.926995,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}]","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-03-07,284139100,132.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Lost in our world, found in another.",John Carter,6.1,2124


## **2. Первичный осмотр данных**

Проведите инвентаризацию датасета. Ответьте на вопросы:
- Сколько строк и столбцов?
- Какие типы данных у каждого столбца? Есть ли некорректные типы?
- Как выглядят базовые статистики числовых признаков?
- Какие проблемы в данных вы видите?


In [53]:
# 1. Размер датасета
print(f"Размер датасета: {df.shape[0]} строк, {df.shape[1]} столбцов")
print()

# 2. Типы данных
print("Типы данных по столбцам:")
print(df.dtypes)
print()

# 3. Базовые статистики числовых признаков
print("Базовые статистики числовых признаков:")
print(df.describe())
print()



Размер датасета: 4803 строк, 20 столбцов

Типы данных по столбцам:
budget                    int64
genres                      str
homepage                    str
id                        int64
keywords                    str
original_language           str
original_title              str
overview                    str
popularity              float64
production_companies        str
production_countries        str
release_date                str
revenue                   int64
runtime                 float64
spoken_languages            str
status                      str
tagline                     str
title                       str
vote_average            float64
vote_count                int64
dtype: object

Базовые статистики числовых признаков:
             budget             id   popularity       revenue      runtime  \
count  4.803000e+03    4803.000000  4803.000000  4.803000e+03  4801.000000   
mean   2.904504e+07   57165.484281    21.492301  8.226064e+07   106.875859   
std  

**Ваши наблюдения по первичному осмотру:**

1. Размер датасета
4803 строки, 20 столбцов

2. Типы данных
- Числовые: budget (int64), id (int64), popularity (float64), revenue (int64), runtime (float64), vote_average (float64), vote_count (int64)
- Строковые (object): genres, homepage, keywords, original_language, original_title, overview, production_companies, production_countries, release_date, spoken_languages, status, tagline, title

Проблемы с типами:
- release_date — строка вместо datetime
- genres, keywords, production_companies, production_countries, spoken_languages — JSON-строки, требуют парсинга

3. Статистики числовых признаков
- Средний бюджет: $29M, медиана: $15M
- Средние сборы: $82M, медиана: $19M
- Средняя длительность: 107 минут
- Средний рейтинг: 6.1/10

## **3. Качество данных**

Исследуйте полноту и чистоту данных:
- В каких столбцах есть пропуски? Какова их доля (в %)?
- Есть ли полные дубликаты строк?
- Сколько фильмов имеют ``budget == 0``? А ``revenue == 0``?
- Что означают нулевые значения в ``budget`` и ``revenue``? Примите решение, что с ними делать.



In [54]:
# Ваш код здесь:
# 1. Проверка пропусков
print("=== Пропуски в столбцах ===")

missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)

missing_df = pd.DataFrame({
    'Пропуски': missing,
    'Доля (%)': missing_pct
})

print(missing_df[missing_df['Пропуски'] > 0])


# 2. Проверка дубликатов
print("\n=== Полные дубликаты строк ===")

duplicates = df.duplicated().sum()

print(f"Количество полных дубликатов: {duplicates}")


# 3. Нулевые значения budget и revenue
print("\n=== Нулевые значения ===")

budget_zero = (df['budget'] == 0).sum()
revenue_zero = (df['revenue'] == 0).sum()

print(
    f"Фильмов с budget == 0: "
    f"{budget_zero} ({budget_zero / len(df) * 100:.2f}%)"
)

print(
    f"Фильмов с revenue == 0: "
    f"{revenue_zero} ({revenue_zero / len(df) * 100:.2f}%)"
)


=== Пропуски в столбцах ===
              Пропуски  Доля (%)
homepage          3091     64.36
overview             3      0.06
release_date         1      0.02
runtime              2      0.04
tagline            844     17.57

=== Полные дубликаты строк ===
Количество полных дубликатов: 0

=== Нулевые значения ===
Фильмов с budget == 0: 1037 (21.59%)
Фильмов с revenue == 0: 1427 (29.71%)


**Найденные проблемы и ваши наблюдения:**

- Пропуски: homepage (64%), tagline (18%), release_date (1 фильм), runtime (2 фильма), overview (3 фильма)
- Нулевые значения: budget=0 у 1037 фильмов (22%), revenue=0 у 1427 фильмов (30%)
- runtime=0 у некоторых фильмов (некорректно)
- vote_average=0 у некоторых фильмов (подозрительно)
- Дубликатов строк нет
- При анализе финансовых показателей фильмы с budget == 0 или revenue == 0 следует исключить.

## **4. Очистка и подготовка данных**

Подготовьте данные к анализу.

**Обязательные шаги:**
1. Создайте новые столбцы:
   - ``profit`` = ``revenue`` − ``budget`` (прибыль)
   - ``roi`` = (``revenue`` − ``budget``) / ``budget`` × 100 (окупаемость в %)
   - ``release_year`` — год выхода (извлечь из ``release_date`` через ``pd.to_datetime().dt.year``)
2. Обработайте пропуски в ``runtime`` (если есть)



In [ ]:
# 1. Создание новых столбцов
# profit = revenue - budget (прибыль)
df['profit'] = df['revenue'] - df['budget']

# roi = (revenue - budget) / budget * 100 (окупаемость в %)
# Для фильмов с budget == 0 ROI будет NaN
df['roi'] = (df['revenue'] - df['budget']) / df['budget'] * 100

# release_year — год выхода из release_date
df['release_date'] = pd.to_datetime(df['release_date'])
df['release_year'] = df['release_date'].dt.year

# 2. Обработка пропусков в runtime
# Заполняем медианным значением
runtime_median = df['runtime'].median()
df['runtime'] = df['runtime'].fillna(runtime_median)

# Проверка результата
print("Новые столбцы созданы:")
print(df[['profit', 'roi', 'release_year', 'runtime']].head())
print(f"\nПропуски в runtime после обработки: {df['runtime'].isnull().sum()}")

Новые столбцы созданы:
       profit          roi  release_year  runtime
0  2550965087  1076.356577        2009.0    162.0
1   661000000   220.333333        2007.0    169.0
2   635674609   259.459024        2015.0    148.0
3   834939099   333.975640        2012.0    165.0
4    24139100     9.284269        2012.0    132.0

Пропуски в runtime после обработки: 0


**Опишите, что вы сделали и почему:**

- Созданы новые признаки profit, roi и release_year. 
    - profit рассчитывается как разница между выручкой и бюджетом, 
    - roi — как процент окупаемости относительно бюджета, 
    - release_year извлечён из даты выхода фильма для дальнейшего анализа по годам.
- Пропуски в runtime заполнены медианным значением, потому что медиана меньше зависит от выбросов и лучше подходит для признака с возможными аномальными значениями.
- После обработки пропусков в runtime не осталось.

## **5. Анализ распределений**

Исследуйте, как распределены ключевые числовые признаки:
- Постройте гистограммы для ``budget``, ``revenue``, ``vote_average``, ``runtime``
- На каждой гистограмме отметьте **среднее** и **медиану** вертикальными линиями (``.axvline()``)
- Сравните среднее и медиану — сильно ли они различаются? О чём это говорит?
- Постройте boxplot для поиска выбросов



In [56]:
# Ваш код здесь:

**Какие закономерности вы видите? Какие распределения скошены? Где выбросы?**

> *Напишите здесь*

## **6. Исследовательские вопросы**

Ответьте на **минимум 5 из 6** вопросов ниже. Для каждого вопроса:
- Напишите код для получения ответа
- Постройте подходящую визуализацию
- Сформулируйте краткий вывод

### Вопрос 1. Какие жанры представлены чаще всего?

Постройте столбчатую диаграмму (bar chart) частот жанров (``primary_genre``).

In [57]:
# Ваш код здесь

**Вывод:**

> *Напишите здесь*

### Вопрос 2. Какие жанры приносят наибольшие средние сборы?

Рассчитайте средний ``revenue`` по жанрам (через ``.groupby()`` + ``.mean()``) и визуализируйте результат.

In [58]:
# Ваш код здесь

**Вывод:**

> *Напишите здесь*

### Вопрос 3. Как менялись сборы и бюджеты по годам?

Сгруппируйте данные по ``release_year`` через ``.groupby()`` + ``.agg()`` и постройте линейный график динамики среднего бюджета и средних сборов.

In [59]:
# Ваш код здесь

**Вывод:**

> *Напишите здесь*

### Вопрос 4. Какие фильмы имеют наибольший ROI (окупаемость)?

Найдите топ-10 фильмов с наибольшим ROI (через `.sort_values()`). Какой жанр преобладает?

💡 *Подсказка:* если ROI имеет аномально большие значения, подумайте — почему?

In [60]:
# Ваш код здесь

**Вывод:**

> *Напишите здесь*

### Вопрос 5. Как распределяются оценки по языку оригинала?

Сравните распределения ``vote_average`` для англоязычных (``original_language == 'en'``) и неанглоязычных фильмов. Постройте гистограммы обеих групп на одном графике.

In [61]:
# Ваш код здесь

**Вывод:**

> *Напишите здесь*

### Вопрос 6. Как длительность фильма связана с оценкой?

Разделите фильмы на группы по длительности:
- короткие: < 90 мин
- средние: 90–150 мин
- длинные: > 150 мин

Сравните средние оценки в каждой группе.

In [62]:
# Ваш код здесь

**Вывод:**

> *Напишите здесь*

---
# **Часть 2. Корреляционный анализ**
---

## **7. Корреляционная матрица**

Постройте корреляционную матрицу для числовых признаков:
``budget``, ``revenue``, ``vote_average``, ``vote_count``, ``runtime``, ``popularity``.

- Рассчитайте коэффициент корреляции, обоснуйте выбор подхода
- Визуализируйте через ``sns.heatmap()`` с аннотациями (``annot=True``)
- Какие пары признаков имеют **сильную** корреляцию (|r| > 0.7)?
- Какие признаки сильнее всего коррелируют с ``revenue``?


In [63]:
# Ваш код здесь:

**Интерпретация корреляционной матрицы:**

> *Напишите здесь: самые сильные связи, их направление и возможные объяснения*

## **8. Визуализация парных связей**

Постройте scatter-графики (диаграммы рассеяния) для ключевых пар:

1. ``budget`` vs ``revenue`` — как бюджет связан со сборами?
2. ``vote_average`` vs ``revenue`` — высокий рейтинг гарантирует сборы?
3. ``vote_count`` vs ``revenue`` — что сильнее связано со сборами: рейтинг или количество голосов?

Для каждого графика добавьте заголовок и подписи осей.

💡 Инструменты: ``plt.scatter()``, ``sns.scatterplot()``, ``plt.xlabel()``, ``plt.ylabel()``, ``plt.title()``

In [64]:
# Ваш код здесь:

**Какие связи вы наблюдаете? Линейные ли они? Есть ли выбросы?**

> *Напишите здесь*

---
# **Часть 3. Проверка статистических гипотез**
---

## **9. Доверительный интервал для средней оценки**

Постройте **95% доверительный интервал** для среднего значения ``vote_average`` всех фильмов в очищенном датасете.

**Шаги:**
1. Рассчитайте выборочное среднее и стандартное отклонение (``.std(ddof=1)``)
2. Рассчитайте стандартную ошибку среднего
3. Найдите критическое значение t-распределения через ``stats.t.ppf()``
4. Вычислите погрешность и границы интервала
5. Проверьте результат через ``stats.t.interval()``

**Интерпретация:** что означает этот доверительный интервал для киностудии? Какую среднюю оценку можно ожидать от фильма?

In [65]:
# Ваш код здесь: ала

**Интерпретация доверительного интервала:**

> *Напишите здесь*

## **10. Гипотезы**

> ✏️ Для получения зачёта выберите как **минимум одну** гипотезу на выбор из двух, представленных в разделе 10.

### **Гипотеза 1. Отличаются ли оценки англоязычных и неанглоязычных фильмов?**

Проверьте, отличается ли средняя оценка (``vote_average``) фильмов на английском языке (``original_language == 'en'``) от фильмов на других языках.

**Шаги:**
1. Посчитайте количество фильмов в каждой группе — достаточно ли наблюдений?
2. Постройте гистограммы обеих групп на одном графике (с ``alpha``)
3. Сформулируйте H₀ и H₁
4. Проведите двухвыборочный тест
5. Рассчитайте размер эффекта (Cohen's d) с помощью ``np.sqrt()``
6. Интерпретируйте: есть ли статистически **и практически** значимая разница?

💡 **Формула Cohen's d:**

d = (x̄₁ − x̄₂) / s_pooled,  где  s_pooled = √((s₁² + s₂²) / 2)

In [66]:
# Ваш код здесь:

**Интерпретация:**

> *H₀: ...*
>
> *H₁: ...*
>
> *t-статистика = ..., p-value = ...*
>
> *Cohen's d = ...*
>
> *Вывод: ...*

### **Гипотеза 2. Дорогие и дешёвые фильмы — одинаковый ли ROI?**

Разделите фильмы на две группы по **медиане бюджета**:
- «Дорогие» — бюджет строго выше медианы
- «Дешёвые» — бюджет меньше или равен медиане

Проверьте, отличается ли средний ``roi`` между этими группами.

💡 *Подсказка:* если в данных есть экстремальные значения ROI, можно отфильтровать фильмы с ``|roi| > 10000%`` — это аномалии (фильмы с микробюджетом), которые искажают анализ.

**Шаги:**
1. Рассчитайте медиану бюджета
2. Создайте два подмножества
3. Сформулируйте H₀ и H₁
4. Проведите двухвыборочный тест
5. Рассчитайте Cohen's d
6. Интерпретируйте: стоит ли делать дорогие фильмы с точки зрения окупаемости?

In [67]:
# Ваш код здесь:

**Интерпретация:**

> *H₀: ...*
>
> *H₁: ...*
>
> *Результат теста: ...*
>
> *Рекомендация для киностудии: ...*

## **11. Свободный анализ**

Выберите **одну** из задач ниже и проведите самостоятельный анализ.

**Вариант A.** Сравните средний ``vote_average`` фильмов двух жанров (например, Action vs Drama). Отличаются ли оценки статистически значимо?

**Вариант B.** Сравните средний ``revenue`` фильмов до 2000 года и после 2000 года. Выросли ли сборы?

**Вариант C.** Придумайте свою гипотезу на основе данных и проверьте её.

Для выбранной задачи:
- Сформулируйте H₀ и H₁
- Постройте доверительный интервал для одной из групп
- Проведите статистический тест
- Интерпретируйте результат

In [68]:
# Ваш код здесь

In [ ]:
# Анализ распределений ключевых числовых признаков

fig, axes = plt.subplots(2, 2, figsize=(14, 10))

# Список признаков для анализа
features = ['budget', 'revenue', 'vote_average', 'runtime']
titles = ['Budget', 'Revenue', 'Vote Average', 'Runtime']

for idx, (col, title) in enumerate(zip(features, titles)):
    ax = axes[idx // 2, idx % 2]
    
    # Гистограмма
    sns.histplot(df[col], bins=50, kde=False, ax=ax, color='skyblue', edgecolor='black')
    
    # Среднее и медиана
    mean_val = df[col].mean()
    median_val = df[col].median()
    
    ax.axvline(mean_val, color='red', linestyle='--', linewidth=2, label=f'Среднее: {mean_val:.2f}')
    ax.axvline(median_val, color='green', linestyle='-', linewidth=2, label=f'Медиана: {median_val:.2f}')
    
    ax.set_title(f'Распределение {title}', fontsize=14)
    ax.set_xlabel(col)
    ax.set_ylabel('Частота')
    ax.legend()

plt.tight_layout()
plt.show()

# Вывод статистик для сравнения
print("=== Сравнение среднего и медианы ===")
print(f"{'Признак':<15} {'Среднее':>15} {'Медиана':>15} {'Разница':>15}")
print("-" * 60)
for col in features:
    mean_val = df[col].mean()
    median_val = df[col].median()
    diff = abs(mean_val - median_val)
    print(f"{col:<15} {mean_val:>15.2f} {median_val:>15.2f} {diff:>15.2f}")

# Boxplot для поиска выбросов
plt.figure(figsize=(14, 6))
sns.boxplot(data=df[features])
plt.title('Boxplot для поиска выбросов', fontsize=14)
plt.ylabel('Значение')
plt.xlabel('Признак')
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

---
# **Итоговые выводы**
---

Сформулируйте итоговый отчёт по результатам комплексного анализа.

**1. Качество данных:**
- Какие основные проблемы обнаружены?
- Как вы их решили?
- Что рекомендуете улучшить при сборе данных?

**2. Ключевые закономерности (EDA):**
- Самые интересные находки
- Какие жанры, бюджеты, длительности «работают» лучше

**3. Корреляционный анализ:**
- Какие признаки сильнее всего связаны с кассовыми сборами?
- Какие связи оказались неожиданными?

**4. Статистические выводы:**
- Результаты проверки гипотез
- Практическая значимость обнаруженных различий

**5. Рекомендации для киностудии:**
- В какие фильмы стоит инвестировать (жанр, бюджет, длительность)?
- Какие дополнительные гипотезы стоит проверить?

**Ваши итоговые выводы:**

> *Напишите здесь*